langroid/agent/batch.py

```diff
@@ -19,66 +22,161 @@
+        drop_exceptions: bool: Whether to replace exceptions with outputs of None
 
     Returns:
-        List[Any]: list of final results
+        list[Any]: list of final results
     """
-
     inputs = [input_map(item) for item in items]
 
-    async def _do_task(input: str | ChatDocument, i: int) -> Any:
-        task_i = task.clone(i)
+    async def _do_task(input: str | ChatDocument, i: int) -> Optional[ChatDocument]:
+        task_i = gen_task(i)
         if task_i.agent.llm is not None:
             task_i.agent.llm.set_stream(False)
         task_i.agent.config.show_stats = False
 
         result = await task_i.run_async(input, turns=turns)
-        return output_map(result)
+        return result
 
-    async def _do_all() -> List[Any]:
-        with quiet_mode(not settings.debug), SuppressLoggerWarnings():
-            if sequential:
-                results = []
-                for i, input in enumerate(inputs):
+    async def _do_all(inputs: list[str | ChatDocument]) -> list[U]:
+        if sequential:
+            results: list[Optional[ChatDocument | BaseException]] = []
+            for i, input in enumerate(inputs):
+                if drop_exceptions:
+                    try:
+                        result: Optional[ChatDocument | BaseException] = await _do_task(
+                            input, i
+                        )
+                    except BaseException as e:
+                        result = e
+                else:
                     result = await _do_task(input, i)
-                    results.append(result)
-                return results
-            return await asyncio.gather(
-                *(_do_task(input, i) for i, input in enumerate(inputs))
+                results.append(result)
+        else:
+            results = await asyncio.gather(
+                *(_do_task(input, i) for i, input in enumerate(inputs)),
+                return_exceptions=drop_exceptions,
             )
 
+        if drop_exceptions:
+            non_exception_results: list[Optional[ChatDocument]] = [
+                None if isinstance(r, BaseException) else r for r in results
+            ]
+        else:
+            non_exception_results: list[Optional[ChatDocument]] = results  # type: ignore
+
+        return list(map(output_map, non_exception_results))
+
     # show rich console spinner
+    status: Callable[[str], Status] | type[nullcontext[str]] = (
+        console.status if console_status else nullcontext  # type: ignore
+    )
 
-    n = len(items)
-    with console.status(f"[bold green]Running {n} copies of {task.name}..."):
-        results = asyncio.run(_do_all())
+    with quiet_mode(not settings.debug), SuppressLoggerWarnings():
+        if batch_size is None:
+            msg = (
+                f"[bold green]Running {len(items)} tasks:"
+                if message is None
+                else message
+            )
+            with status(msg):
```

Comment:
Not sure, but if `status = nullcontext` then the `[bold green]` in the `msg` won't work as intended and show up as a string?

Reply:
No, in that case nothing is shown at all